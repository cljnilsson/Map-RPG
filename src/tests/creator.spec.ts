import { expect, test, beforeEach, vi } from "vitest";
import { render, screen, fireEvent } from "@testing-library/svelte";
import Creator from "$lib/features/creator/creator.svelte";
import CreatorStat from "$lib/features/creator/creatorStat.svelte";

test("Creator renders with correct elements", () => {
	const {container} = render(Creator);
	expect(container).toMatchSnapshot();
});

test("Stat '+' buttons are disabled when no points are left", () => {
	const min: number = 3;
	const max: number = 20;
	const totalMax: number = 30;

	render(CreatorStat, {props: {
		min , max, total: 15, totalMax, stat: 7, name: "Vit", totalLeft: 0
	}});

	const plusButtons = screen.getAllByRole("button", { name: "+" });
	expect(plusButtons.length).toBe(1); // One for each stat
	plusButtons.forEach((btn) => expect(btn).toBeDisabled());
});

test("Stat '-' buttons are enabled and work", async () => {
	const min: number = 3;
	const max: number = 20;
	const totalMax: number = 30;

	render(CreatorStat, {props: {
		min , max, total: 15, totalMax, stat: 7, name: "Vit", totalLeft: 0
	}});

	const minusButtons = screen.getAllByRole("button", { name: "-" });
	expect(minusButtons.length).toBe(1);

	// Get the "Vit" row and its current value
	const vitRow = screen.getByText("Vit").closest(".row")!;
	const valueDiv = vitRow.querySelector(".col-auto")!;
	const initialValue = Number(valueDiv.textContent);

	await fireEvent.click(minusButtons[0]); // Decrease Vit
	const newValue = Number(valueDiv.textContent);

	expect(newValue).toBeLessThan(initialValue);
});

test("Input fields for Name and Age are editable", async () => {
	render(Creator);
	const nameInput = screen.getByPlaceholderText("Name") as HTMLInputElement;
	const ageInput = screen.getByPlaceholderText("Age") as HTMLInputElement;

	await fireEvent.input(nameInput, { target: { value: "Hero123" } });
	await fireEvent.input(ageInput, { target: { value: "27" } });

	expect(nameInput.value).toBe("Hero123");
	expect(ageInput.value).toBe("27");
});

test("Character tab is selected by default", () => {
	render(Creator);
	const characterTab = screen.getByRole("button", { name: "Character" });
	expect(characterTab.className).toContain("activeSelection");
});
