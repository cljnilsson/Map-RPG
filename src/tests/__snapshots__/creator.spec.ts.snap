// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Creator renders with correct elements 1`] = `
<div>
  <div
    class="row justify-content-center px-3 mt-5 position-relative"
  >
    <div
      class="col-xl-10 col-md-12 wrapper svelte-17t8ei6"
    >
      <div
        class="c-header"
      >
        <h3
          class="text-center"
        >
          Create your character!
        </h3>
        <!---->
      </div>
       
      <div
        class="c-body"
      >
        <div
          class="row justify-content-center my-5"
        >
          <div
            class="col-12 position-relative"
          >
            <div
              class="char-menu svelte-17t8ei6"
            >
              <div>
                <button
                  class="btn svelte-17t8ei6 activeSelection"
                >
                  <h5>
                    Character
                  </h5>
                </button>
              </div>
               
              <div>
                <button
                  class="btn svelte-17t8ei6"
                >
                  <h5>
                    Class
                  </h5>
                </button>
              </div>
               
              <div>
                <button
                  class="btn svelte-17t8ei6"
                >
                  <h5>
                    Faith
                  </h5>
                </button>
              </div>
               
              <div>
                <button
                  class="btn svelte-17t8ei6"
                >
                  <h5>
                    Image
                  </h5>
                </button>
              </div>
            </div>
             
            <div
              class="main-content offset"
            >
              <div
                class="row justify-content-center"
              >
                <div
                  class="col-auto"
                >
                  <div
                    class="c-avatar text-center"
                  >
                    <img
                      alt="avatar"
                      height="200px"
                      loading="lazy"
                      src="https://cdn4.iconfinder.com/data/icons/small-n-flat/24/user-alt-512.png"
                      width="200px"
                    />
                  </div>
                </div>
              </div>
               
              <div
                class="row mt-5"
              >
                <div
                  class="col text-center"
                >
                  Suggested stats: 
                  <strong>
                    Strength
                  </strong>
                  , 
                  <strong>
                    Vitality
                  </strong>
                  
                  <!---->
                </div>
              </div>
               
              <div
                class="row mb-5"
              >
                <div
                  class="col text-center"
                >
                  Points left: 0
                </div>
              </div>
               
              <div
                class="c-stats"
              >
                <div
                  class="row justify-content-center align-items-center my-2 stat-Str"
                >
                  <div
                    class="col-xl-2 col-md-4"
                  >
                    Str
                  </div>
                   
                  <div
                    class="col-auto"
                  >
                    6
                  </div>
                   
                  <div
                    class="col-xl-2 col-md-6 text-center"
                  >
                    <button
                      class="btn btn-outline-dark"
                      disabled=""
                    >
                      +
                    </button>
                     
                    <button
                      class="btn btn-outline-dark"
                    >
                      -
                    </button>
                  </div>
                </div>
                <!---->
                 
                <div
                  class="row justify-content-center align-items-center my-2 stat-Dex"
                >
                  <div
                    class="col-xl-2 col-md-4"
                  >
                    Dex
                  </div>
                   
                  <div
                    class="col-auto"
                  >
                    6
                  </div>
                   
                  <div
                    class="col-xl-2 col-md-6 text-center"
                  >
                    <button
                      class="btn btn-outline-dark"
                      disabled=""
                    >
                      +
                    </button>
                     
                    <button
                      class="btn btn-outline-dark"
                    >
                      -
                    </button>
                  </div>
                </div>
                <!---->
                 
                <div
                  class="row justify-content-center align-items-center my-2 stat-Int"
                >
                  <div
                    class="col-xl-2 col-md-4"
                  >
                    Int
                  </div>
                   
                  <div
                    class="col-auto"
                  >
                    6
                  </div>
                   
                  <div
                    class="col-xl-2 col-md-6 text-center"
                  >
                    <button
                      class="btn btn-outline-dark"
                      disabled=""
                    >
                      +
                    </button>
                     
                    <button
                      class="btn btn-outline-dark"
                    >
                      -
                    </button>
                  </div>
                </div>
                <!---->
                 
                <div
                  class="row justify-content-center align-items-center my-2 stat-Vit"
                >
                  <div
                    class="col-xl-2 col-md-4"
                  >
                    Vit
                  </div>
                   
                  <div
                    class="col-auto"
                  >
                    6
                  </div>
                   
                  <div
                    class="col-xl-2 col-md-6 text-center"
                  >
                    <button
                      class="btn btn-outline-dark"
                      disabled=""
                    >
                      +
                    </button>
                     
                    <button
                      class="btn btn-outline-dark"
                    >
                      -
                    </button>
                  </div>
                </div>
                <!---->
                 
                <div
                  class="row justify-content-center align-items-center my-2 stat-Charisma"
                >
                  <div
                    class="col-xl-2 col-md-4"
                  >
                    Charisma
                  </div>
                   
                  <div
                    class="col-auto"
                  >
                    6
                  </div>
                   
                  <div
                    class="col-xl-2 col-md-6 text-center"
                  >
                    <button
                      class="btn btn-outline-dark"
                      disabled=""
                    >
                      +
                    </button>
                     
                    <button
                      class="btn btn-outline-dark"
                    >
                      -
                    </button>
                  </div>
                </div>
                <!---->
              </div>
              <!---->
            </div>
          </div>
        </div>
         
        <div
          class="c-footer"
        >
          <div
            class="row justify-content-center my-5"
          >
            <div
              class="col-xl-4 col-md-10"
            >
              <input
                class="form-control"
                placeholder="Name"
                type="text"
              />
            </div>
             
            <div
              class="col-xl-2 col-md-10"
            >
              <input
                class="form-control"
                placeholder="Age"
                type="number"
              />
            </div>
          </div>
          <!---->
        </div>
      </div>
       
      <div
        class="text-center"
      >
        <button
          class="btn btn-lg btn-primary"
          disabled=""
        >
          Create!
        </button>
      </div>
    </div>
  </div>
  
</div>
`;
